### PR TITLE
Support static type checking with `mypy`

### LINKED ISSUE #301: Proposal to deprecate python 3.8
Python 3.8 does not support the use of native `tuple` and `list` for type annotations.

Maintaining support python 3.8 means carrying around `from typing import List, Tuple` statements.

Something against discontinuing support for 3.8, @JannisNe , @robertdstein ?

### PR DESCRIPTION
**Pull request by mlincett**
This PR provides a minimal set of change to let a `mypy` run pass on the current codebase. The primary reason for this is to allow the use of mypy to test improvements and refactors of the existing code.

Note that `mypy` (in my understanding) checks for type consistency only functions that have a typed return value, and optionally all functions with at least one type annotation even if untyped. It also does more general QC, function signature-based syntax checking. Notably it does not like variable name re-definitions.

This PR also deprecates python 3.8 and closes #301 so we can use native `tuple` and `list` type annotations.

To summarise the changes, they mostly fall in one of these categories:
- type annotations for the `subclasses` dictionary we use in several `flarestack.core` functionalities;
- style improvements while I was at it (f-strings, alphabetical order of imports etc.)
- type annotation with `Optional[T]` for variables that can be of type `T` or `None`; on the long term, conditional statements based on `None` values should be likely deprecated;
- commenting out of code affected by non-obvious syntax errors: there are several code blocks under `misc` or `utils` that are no longer compatible with the current flarestack core API and that are evidently untested (otherwise their test would fail);
- avoid variable redefinitions such as `x = np.array(x)` since a variable should not change its type; this may seem a bit too strict but on the long term it improves clarity.

### CODE REVIEW

File: flarestack/core/llh.py [hunk 1]
```diff
@@ -1394,13 +1394,18 @@ def find_significant_events(self, coincident_data, source):
     return FlareLLH(season, sources, llh_dict)
 
 
+"""
+# NOTE: `mypy` allowed to find some deprecations in the following __main__ block that were never noticed before, because the block is not part of any testing routine. It should be checked whether equivalent code is covered by an unit test, before removing the (now commented) block altogether.
+
 if __name__ == "__main__":
```
Comment (robertdstein) [inline flarestack/core/llh.py:1400]: I think delete this __main__ bit

**Review by robertdstein — APPROVED** [review]
Approved pending 1 comment

**Comment by robertdstein** [comment]
I updated the CI. It's under settings, then Branch protection rules. I removed python 3.8, and added python 3.11, as a required check.

**Comment by mlincett** [comment]
> I updated the CI. It's under settings, then Branch protection rules. I removed python 3.8, and added python 3.11, as a required check.

Looked for it but could not find it. Apparently I don't have high enough privileges (I guess).

**Pull request merged**